File: examples/crud-item-list/src/androidTest/java/com/gurunars/crud_item_list/example/ActivitySortableTest.java
package com.gurunars.crud_item_list.example;

import android.content.pm.ActivityInfo;
import android.support.test.espresso.ViewInteraction;
import android.support.test.filters.LargeTest;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static android.support.test.InstrumentationRegistry.getInstrumentation;
import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.action.ViewActions.longClick;
import static android.support.test.espresso.assertion.ViewAssertions.doesNotExist;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.isEnabled;
import static android.support.test.espresso.matcher.ViewMatchers.withContentDescription;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static com.gurunars.test_utils.Helpers.nthChildOf;
import static org.hamcrest.core.IsNot.not;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class ActivitySortableTest {

    private void restart() {
        mActivityRule.getActivity().setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
        mActivityRule.getActivity().setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
    }

    @Rule
    public ActivityTestRule<ActivityMain> mActivityRule = new ActivityTestRule<>(
            ActivityMain.class);

    private void validateEnabled(int id) {
        onView(withId(id)).check(matches(isEnabled()));
    }

    private void validateDisabled(int id) {
        onView(withId(id)).check(matches(not(isEnabled())));
    }

    private void validateDoesNotExist(int id) {
        onView(withId(id)).check(doesNotExist());
    }

    private ViewInteraction atIndex(int id) {
        return onView(nthChildOf(withId(R.id.recyclerView), id));
    }

    @Test
    public void selectingOne_shouldOpenContextualMenu() {
        validateDoesNotExist(R.id.moveUp);
        validateDoesNotExist(R.id.moveDown);
        validateDoesNotExist(R.id.edit);
        validateDoesNotExist(R.id.delete);
        validateDoesNotExist(R.id.selectAll);
        atIndex(1).perform(longClick());
        restart();
        validateEnabled(R.id.moveUp);
        validateEnabled(R.id.moveDown);
        validateEnabled(R.id.edit);
        validateEnabled(R.id.delete);
        validateEnabled(R.id.selectAll);
    }

    @Test
    public void deselectingLast_shouldCloseContextualMenu() {
        atIndex(0).perform(longClick());
        restart();
        atIndex(0).perform(click());
        restart();
        validateDoesNotExist(R.id.moveUp);
        validateDoesNotExist(R.id.moveDown);
        validateDoesNotExist(R.id.edit);
        validateDoesNotExist(R.id.delete);
        validateDoesNotExist(R.id.selectAll);
    }

    @Test
    public void clickingCross_shouldCloseContextualMenu() {
        atIndex(0).perform(longClick());
        restart();
        onView(withId(R.id.openFab)).perform(click());
        restart();
        validateDoesNotExist(R.id.moveUp);
        validateDoesNotExist(R.id.moveDown);
        validateDoesNotExist(R.id.edit);
        validateDoesNotExist(R.id.delete);
        validateDoesNotExist(R.id.selectAll);
    }

    @Test
    public void selectingTopItem_shouldDisableMoveUp() {
        atIndex(0).perform(longClick());
        restart();
        validateDisabled(R.id.moveUp);
        validateEnabled(R.id.moveDown);
        validateEnabled(R.id.edit);
        validateEnabled(R.id.delete);
        validateEnabled(R.id.selectAll);
    }

    @Test
    public void selectingBottomItem_shouldDisableMoveDown() {
        atIndex(3).perform(longClick());
        restart();
        validateEnabled(R.id.moveUp);
        validateDisabled(R.id.moveDown);
        validateEnabled(R.id.edit);
        validateEnabled(R.id.delete);
        validateEnabled(R.id.selectAll);
    }

    @Test
    public void selectingUnsolidChunk_shouldDisableMoveAndEdit() {
        atIndex(0).perform(longClick());
        restart();
        atIndex(3).perform(click());
        restart();
        validateDisabled(R.id.moveUp);
        validateDisabled(R.id.moveDown);
        validateDisabled(R.id.edit);
        validateEnabled(R.id.delete);
        validateEnabled(R.id.selectAll);
    }

    @Test
    public void selectingSolidChunk_shouldEnableMoveButDisableEdit() {
        atIndex(1).perform(longClick());
        restart();
        atIndex(2).perform(click());
        restart();
        validateEnabled(R.id.moveUp);
        validateEnabled(R.id.moveDown);
        validateDisabled(R.id.edit);
        validateEnabled(R.id.delete);
        validateEnabled(R.id.selectAll);
    }

    @Test
    public void movingSelectionToTop_shouldDisableMoveUp() {
        atIndex(1).perform(longClick());
        restart();
        onView(withId(R.id.moveUp)).perform(click());
        restart();
        validateDisabled(R.id.moveUp);
        validateEnabled(R.id.moveDown);
        validateEnabled(R.id.edit);
        validateEnabled(R.id.delete);
        validateEnabled(R.id.selectAll);
    }

    @Test
    public void movingSelectionToBottom_shouldDisableMoveDown() {
        atIndex(2).perform(longClick());
        restart();
        onView(withId(R.id.moveDown)).perform(click());
        restart();
        validateEnabled(R.id.moveUp);
        validateDisabled(R.id.moveDown);
        validateEnabled(R.id.edit);
        validateEnabled(R.id.delete);
        validateEnabled(R.id.selectAll);
    }

    @Test
    public void editingItem_shouldIncrementVersion() {
        atIndex(3).perform(longClick());
        restart();
        onView(withId(R.id.edit)).perform(click());
        restart();
        atIndex(3).check(matches(withText("#4{WOLF @ 2|true}")));
    }

    @Test
    public void creatingItem_shouldIncrementItemListSize() {
        onView(withId(R.id.openFab)).perform(click());
        restart();
        onView(withId(R.id.lion)).perform(click());
        restart();
        atIndex(4).check(matches(withText("#5{LION @ 1|false}")));
    }

    @Test
    public void removingAllItems_shouldShowSpecialDefaultLayout() {
        atIndex(1).perform(longClick());
        restart();
        onView(withId(R.id.selectAll)).perform(click());
        onView(withId(R.id.delete)).perform(click());
        onView(withId(R.id.noItemsLabel)).check(matches(withText("No items at all")));
    }

    @Test
    public void leftAndRightHandSwitch() {
        atIndex(3).perform(longClick());
        restart();
        onView(withId(R.id.contextualMenu)).check(matches(withContentDescription("RIGHT HANDED")));
        openActionBarOverflowOrOptionsMenu(getInstrumentation().getTargetContext());
        onView(withText("Left handed")).perform(click());
        restart();
        onView(withId(R.id.contextualMenu)).check(matches(withContentDescription("LEFT HANDED")));
        openActionBarOverflowOrOptionsMenu(getInstrumentation().getTargetContext());
        onView(withText("Right handed")).perform(click());
        restart();
        onView(withId(R.id.contextualMenu)).check(matches(withContentDescription("RIGHT HANDED")));
    }

    @Test
    public void testSelectMoveUpAndReset() {
        atIndex(1).perform(longClick());
        onView(withId(R.id.moveUp)).perform(click());
        onView(withId(R.id.reset)).perform(click());
        atIndex(0).check(matches(withText("#1{LION @ 1|false}")));
        atIndex(1).check(matches(withText("#2{TIGER @ 1|true}")));
    }

    @Test
    public void testSelectMoveDownAndReset() {
        atIndex(2).perform(longClick());
        onView(withId(R.id.moveDown)).perform(click());
        onView(withId(R.id.reset)).perform(click());
        atIndex(2).check(matches(withText("#3{MONKEY @ 1|true}")));
        atIndex(3).check(matches(withText("#4{WOLF @ 1|false}")));
    }

    @Before
    public void before() {
        onView(withId(R.id.reset)).perform(click());
        onView(withId(R.id.unlock)).perform(click());
    }

    @After
    public void after() {
        onView(withId(R.id.reset)).perform(click());
    }

}

File: examples/crud-item-list/src/androidTest/kotlin/com/gurunars/crud_item_list/example/ActivitySortableTest.kt
package com.gurunars.crud_item_list.example

import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.ViewInteraction
import android.support.test.espresso.action.ViewActions.*
import android.support.test.espresso.assertion.ViewAssertions.matches
import android.support.test.espresso.matcher.ViewMatchers
import android.support.test.espresso.matcher.ViewMatchers.*
import android.support.test.filters.LargeTest
import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import android.view.View
import com.gurunars.test_utils.Helpers.nthChildOf
import com.gurunars.test_utils.rotate
import org.hamcrest.core.Is.`is`
import org.hamcrest.core.IsNot.not
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
@LargeTest
class ActivitySortableTest {

    private fun rotate() {
        mActivityRule.rotate()
    }

    @get:Rule
    var mActivityRule = ActivityTestRule(
        ActivityMain::class.java)

    private fun validateEnabled(id: Int) {
        onView(withId(id)).check(matches(isEnabled()))
    }

    private fun validateDisabled(id: Int) {
        onView(withId(id)).check(matches(not<View>(isEnabled())))
    }

    private fun validateDoesNotExist(id: Int) {
        onView(withId(id)).check(matches(withEffectiveVisibility(ViewMatchers.Visibility.GONE)))
    }

    private fun validateExists(id: Int) {
        onView(withId(id)).check(matches(withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)))
    }

    private fun atIndex(id: Int): ViewInteraction {
        return onView(nthChildOf(withId(R.id.recyclerView), id))
    }

    private fun validateSelection(index: Int, text: String, isSelected: Boolean) {
        onView(nthChildOf(nthChildOf(withId(R.id.recyclerView), index), 1)).check(matches(withText(text)))
        atIndex(index).check(matches(withTagKey(R.id.isSelected, `is`(isSelected))))
    }

    private fun sleep() {
        Thread.sleep(2100)
    }

    @Test
    fun selectingOne_shouldOpenContextualMenu() {
        validateDoesNotExist(R.id.menuPane)
        atIndex(1).perform(longClick())
        rotate()
        validateExists(R.id.menuPane)
        validateEnabled(R.id.edit)
        validateEnabled(R.id.moveUp)
        validateEnabled(R.id.moveDown)
        validateEnabled(R.id.delete)
        validateEnabled(R.id.selectAll)
    }

    @Test
    fun deselectingLast_shouldCloseContextualMenu() {
        validateDoesNotExist(R.id.menuPane)
        atIndex(0).perform(longClick())
        validateExists(R.id.menuPane)
        rotate()
        validateExists(R.id.menuPane)
        atIndex(0).perform(click())
        rotate()
        validateDoesNotExist(R.id.menuPane)
    }

    @Test
    fun clickingTick_shouldCloseContextualMenu() {
        validateDoesNotExist(R.id.menuPane)
        atIndex(0).perform(longClick())
        validateExists(R.id.menuPane)
        rotate()
        validateExists(R.id.menuPane)
        onView(withId(R.id.openFab)).perform(click())
        rotate()
        validateDoesNotExist(R.id.menuPane)
    }

    @Test
    fun selectingTopItem_shouldDisableMoveUp() {
        atIndex(0).perform(longClick())
        rotate()
        validateDisabled(R.id.moveUp)
        validateEnabled(R.id.edit)
        validateEnabled(R.id.moveDown)
        validateEnabled(R.id.delete)
        validateEnabled(R.id.selectAll)
    }

    @Test
    fun selectingBottomItem_shouldDisableMoveDown() {
        atIndex(3).perform(longClick())
        rotate()
        validateEnabled(R.id.moveUp)
        validateEnabled(R.id.edit)
        validateDisabled(R.id.moveDown)
        validateEnabled(R.id.delete)
        validateEnabled(R.id.selectAll)
    }

    @Test
    fun selectingUnsolidChunk_shouldDisableMove() {
        atIndex(0).perform(longClick())
        rotate()
        atIndex(3).perform(click())
        rotate()
        validateDisabled(R.id.moveUp)
        validateDisabled(R.id.moveDown)
        validateDisabled(R.id.edit)
        validateEnabled(R.id.delete)
        validateEnabled(R.id.selectAll)
    }

    @Test
    fun selectingSolidChunk_shouldEnableMove() {
        atIndex(1).perform(longClick())
        rotate()
        atIndex(2).perform(click())
        rotate()
        validateDisabled(R.id.edit)
        validateEnabled(R.id.moveUp)
        validateEnabled(R.id.moveDown)
        validateEnabled(R.id.delete)
        validateEnabled(R.id.selectAll)
    }

    @Test
    fun movingSelectionToTop_shouldDisableMoveUp() {
        atIndex(1).perform(longClick())
        rotate()
        onView(withId(R.id.moveUp)).perform(click())
        rotate()
        validateDisabled(R.id.moveUp)
        validateEnabled(R.id.edit)
        validateEnabled(R.id.moveDown)
        validateEnabled(R.id.delete)
        validateEnabled(R.id.selectAll)
    }

    @Test
    fun movingSelectionToBottom_shouldDisableMoveDown() {
        atIndex(2).perform(longClick())
        rotate()
        onView(withId(R.id.moveDown)).perform(click())
        rotate()
        validateEnabled(R.id.moveUp)
        validateDisabled(R.id.moveDown)
        validateEnabled(R.id.edit)
        validateEnabled(R.id.delete)
        validateEnabled(R.id.selectAll)
    }

    @Test
    fun removingAllItems_shouldShowSpecialDefaultLayout() {
        atIndex(1).perform(longClick())
        rotate()
        onView(withId(R.id.selectAll)).perform(click())
        onView(withId(R.id.delete)).perform(click())
        onView(withId(R.id.noItemsLabel)).check(matches(withText("Empty")))
    }

    @Test
    fun testSelectMoveUpAndReset() {
        atIndex(1).perform(longClick())
        onView(withId(R.id.moveUp)).perform(click())
        onView(withId(R.id.reset)).perform(click())
        sleep()
        validateSelection(0, "#1{LION @ 0}", false)
        validateSelection(1, "#2{TIGER @ 0}", false)
    }

    // TODO: flaky
    @Test
    fun testCopyAndPaste() {
        atIndex(1).perform(longClick())
        atIndex(2).perform(click())
        rotate()
        onView(withId(R.id.copy)).perform(click())
        onView(withId(R.id.paste)).perform(click())
        sleep()
        rotate()
        atIndex(4).perform(click())
        atIndex(5).perform(click())

        validateSelection(1, "#2{TIGER @ 0}", true)
        validateSelection(2, "#3{MONKEY @ 0}", true)
        validateSelection(3, "#5{TIGER @ 0}", false)
        validateSelection(4, "#6{MONKEY @ 0}", true)
        validateSelection(5, "#4{WOLF @ 0}", true)
    }

    @Test
    fun testSelectMoveDownAndReset() {
        atIndex(2).perform(longClick())
        onView(withId(R.id.moveDown)).perform(click())
        onView(withId(R.id.reset)).perform(click())
        sleep()
        validateSelection(2, "#3{MONKEY @ 0}", false)
        validateSelection(3, "#4{WOLF @ 0}", false)
    }

    // TODO: flaky
    @Test
    fun editingItem_shouldIncrementVersion() {
        atIndex(3).perform(longClick())
        onView(withId(R.id.edit)).perform(click())
        onView(withId(R.id.increment)).perform(click())
        rotate()
        onView(withId(R.id.save)).perform(click())
        // Save operation updates the value asynchronously - we need to wait
        sleep()
        rotate()
        validateSelection(3, "#4{WOLF @ 1}", false)
    }

    @Test
    fun creatingItem_shouldIncrementItemListSize() {
        onView(withId(R.id.openFab)).perform(click())
        rotate()
        onView(withTagValue(`is`("LION"))).perform(click())
        onView(withId(R.id.versionValue)).perform(typeText("4"))
        onView(withId(R.id.increment)).perform(click())
        rotate()
        onView(withId(R.id.save)).perform(click())
        sleep()
        validateSelection(4, "#5{LION @ 5}", false)
    }

    @Test
    fun longPressingPlus_shouldOpenContextualMenu() {
        onView(withId(R.id.openFab)).perform(longClick())
        rotate()
        validateExists(R.id.menuPane)
        validateDisabled(R.id.paste)
        validateDisabled(R.id.copy)
        validateDisabled(R.id.edit)
        validateDisabled(R.id.moveUp)
        validateDisabled(R.id.moveDown)
        validateDisabled(R.id.delete)
        validateEnabled(R.id.selectAll)
    }

    @Test
    fun explicitContextualMenu_shouldEnableSelectionViaOneClick() {
        onView(withId(R.id.openFab)).perform(longClick())
        rotate()
        atIndex(2).perform(click())
        validateSelection(1, "#2{TIGER @ 0}", false)
        validateSelection(2, "#3{MONKEY @ 0}", true)
        validateSelection(3, "#4{WOLF @ 0}", false)
    }

    @Before
    fun before() {
        onView(withId(R.id.reset)).perform(click())
        onView(withId(R.id.unlock)).perform(click())
        sleep()
    }

    @After
    fun waitd() {
        //Thread.sleep(100000)
    }

}